### PR TITLE
Npm gitlab url work

### PR DESCRIPTION
**Pull request by vivanov00**
https://github.com/renovatebot/renovate/pull/4693/files/fed6ad93d614aa83f8ba6e39d32daa9253b657e1#r336871955 fixed.
Lint errors fixed.

### CODE REVIEW

File: lib/datasource/index.ts [hunk 1]
```diff
@@ -124,3 +126,51 @@ export function getDigest(
     value
   );
 }
+export function sanitizeSourceUrl(rawSourceUrl) {
```
Comment (rarkins) [inline lib/datasource/index.ts]: This function needs to handle null/undefined and exit early

File: lib/datasource/index.ts [hunk 2]
```diff
@@ -124,3 +126,51 @@ export function getDigest(
...
+  let repoUrlTmp: string;
+  if (!rawSourceUrl.endsWith('.git')) {
+    logger.debug(`${rawSourceUrl} is not a valid git url.`);
```
Comment (rarkins) [inline lib/datasource/index.ts]: Are you sure this is a valid assumption? I would prefer to use `hostedGitInfo` as the first step anyway

File: lib/datasource/index.ts [hunk 3]
```diff
@@ -124,3 +126,51 @@ export function getDigest(
...
+    return null;
+  }
+  const git_info = hostedGitInfo.fromUrl(rawSourceUrl, { noGitPlus: true });
```
Comment (rarkins) [inline lib/datasource/index.ts]: camelCase

File: lib/datasource/index.ts [hunk 4]
```diff
@@ -124,3 +126,51 @@ export function getDigest(
...
+  if (git_info === undefined) {
```
Comment (rarkins) [inline lib/datasource/index.ts]: switch this logic around, e.g. `if (gitInfo) {`

File: lib/datasource/npm/get.ts [hunk 5]
```diff
@@ -174,6 +174,9 @@ export async function getDependency(
         extraBaseUrls,
       });
     }
+    if (sourceUrl === undefined) {
+      sourceUrl = res.repository.url;
```
Comment (rarkins) [inline lib/datasource/npm/get.ts]: I think if we are doing sourceUrl parsing in the datasource/index then we can remove the custom parsing from this npm datasource and just return the raw url.

File: package.json [hunk 6]
```diff
@@ -120,6 +120,7 @@
     "got": "9.6.0",
     "handlebars": "4.4.3",
     "hasha": "5.1.0",
+    "hosted-git-info": "^3.0.2",
```
Comment (rarkins) [inline package.json]: pin this

File: package.json [hunk 7]
```diff
@@ -162,13 +163,13 @@
     "re2": "1.10.4"
   },
   "devDependencies": {
-    "@babel/cli": "7.6.2",
-    "@babel/core": "7.6.2",
-    "@babel/node": "7.6.2",
+    "@babel/cli": "7.6.4",
```
Comment (rarkins) [inline package.json]: Not sure why these are upgraded?

File: lib/util/host-rules.ts [hunk 8]
```diff
@@ -153,6 +153,21 @@ export function find(search: HostRuleSearch) {
   return res;
 }
 
+export function getPlatformByHostOrUrl(url) {
```
Comment (rarkins) [inline lib/util/host-rules.ts]: Is this used?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/index.ts (hunk 2 above)
Comment (vivanov00) [reply lib/datasource/index.ts]: hostedGitInfo requires that the uri end in git, otherwise it just returns undefined.

Comment (rarkins) [reply lib/datasource/index.ts]: That's not the impression that I got from their docs:

image

Comment (vivanov00) [reply lib/datasource/index.ts]: Ah, my mistake, I got a few errors while testing it via a script I wrote to test my custom function.
Will fix.

Comment (vivanov00) [reply lib/datasource/index.ts]: Fixed.

File: lib/datasource/index.ts (hunk 3 above)
Comment (vivanov00) [reply lib/datasource/index.ts]: Fixed.

File: lib/datasource/index.ts (hunk 4 above)
Comment (vivanov00) [reply lib/datasource/index.ts]: Fixed.

File: lib/datasource/npm/get.ts (hunk 5 above)
Comment (vivanov00) [reply lib/datasource/npm/get.ts]: Done.

File: package.json (hunk 6 above)
Comment (vivanov00) [reply package.json]: Done.

File: package.json (hunk 7 above)
Comment (vivanov00) [reply package.json]: fixed.

File: lib/util/host-rules.ts (hunk 8 above)
Comment (vivanov00) [reply lib/util/host-rules.ts]: Yes it does, however I forgot to pull that logic from another fork, here is the current state of affairs:
https://github.com/ocadotechnology/renovate/blob/04fb68896d803688ac6c094edbeebebc05615826/lib/datasource/index.ts#L142
You did mention that you wish for the flow to be something along the lines of:
1) Check if this is a valid public repo, done via 'hosted-git-info'
2) Check if there is a host rule for the relevant hostname.
Fixed.

File: lib/datasource/index.ts [hunk 9]
```diff
@@ -106,6 +108,7 @@ async function fetchReleases(
   }
   const dep = await datasources[datasource].getPkgReleases(config);
   addMetaData(dep, datasource, config.lookupName);
+  dep.sourceUrl = sanitizeSourceUrl(dep.sourceUrl);
```
Comment (rarkins) [inline lib/datasource/index.ts]: This behaviour will result in every result without a sourceUrl now having one that is `null`. It's cleaner if it's simply absent than null. i.e. `sanitizeSourceUrl() === null`, which is not desirable.

Comment (vivanov00) [reply lib/datasource/index.ts]: https://github.com/ocadotechnology/renovate/commit/76559ef421d5d3b69ba6c98cafa0741589456e0d#diff-a367c71b38b3046d1795a93024b87f28
should fix this.

**Pull request closed without merging**